[ONNX] Diagnostic to show all unsupported call_functions

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100451
* #100407

Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

BowenBao committed:
[ONNX] Diagnostic to show all unsupported call_functions

[ghstack-poisoned]

BowenBao committed:
forgot new files on "[ONNX] Diagnostic to show all unsupported call_functions"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Diagnostic to show all unsupported call_functions"


Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

[ghstack-poisoned]

BowenBao committed:
rebase and minor tweak on "[ONNX] Diagnostic to show all unsupported call_functions"


Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -306,3 +309,82 @@ def _create_onnx_friendly_decomposition_table() -> (
 ] = _create_onnx_friendly_decomposition_table()
+
+
+def get_symbolic_function(
```
note: this part is extracting dispatch code from `fx_to_onnxscript.py` to `function_dispatcher.py`. The goal is to make it re-usable for the analysis pass. A proper dispatcher refactor is still required, and is not the scope of this PR.

titaiwangms replied:
#100660 targets the refactor of `function_dispatcher`

titaiwangms commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -204,3 +204,22 @@ def run(self, *args, **kwargs) -> torch.fx.GraphModule:
+
+
+class Analysis(abc.ABC):
+    @_beartype.beartype
```
docstring

titaiwangms commented on torch/onnx/_internal/fx/analysis/unsupported_nodes.py:
```diff
@@ -0,0 +1,80 @@
+    unsupported_op_to_target_mapping: Dict[str, Set[str]]
+
+
+class UnsupportedFxNodesAnalysis(_pass.Analysis):
```
Is this already in use?

BowenBao replied:
It is. It is invoked before calling `export_fx_to_onnxscript`. It is also how the benchmark gathers all missing op info for a model: by analyzing the diagnostics emitted by this analysis.

titaiwangms replied:
I see

titaiwangms commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -306,3 +309,82 @@ def _create_onnx_friendly_decomposition_table() -> (
+                # TODO: reduce number of explicit initializations.
+                # TODO: Log location, stack.
+                diagnostic = diagnostics.UnsupportedFxNodeDiagnostic(
+                    diagnostics.rules.no_symbolic_function_for_call_function,
```
Not sure if I understand this correctly. This PR registers two rules, but only `no_symbolic_function_for_call_function` is actually guarding errors. It seems the other one is prepared for some kind of full analysis, and it's not done yet? Which means it's not an actual rule to apply on converter code. It's rather a "wrapper rule" on the rules?

titaiwangms replied:
Why do we need another layer to reveal unsupported node?

titaiwangms commented:
@pytorchbot merge

titaiwangms commented:
@pytorchbot rebase

BowenBao committed:
Update on "[ONNX] Diagnostic to show all unsupported call_functions"


Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Diagnostic to show all unsupported call_functions"


Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Diagnostic to show all unsupported call_functions"


Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
